File: pipelines/inference.py
import re
import subprocess

from pipelines.config import MMPROJ_PATH, MODEL_PATH

_KV_LINE = re.compile(r"^([A-Z_]+):\s*(.*?)\s*$")


def run_vlm(image_path: str, prompt_path: str, n_predict: int = 150) -> dict:
    """Run the InternVL GGUF model on one image with one prompt file via llama.cpp.

    Returns a dict of the KEY: VALUE fields the model printed (e.g.
    {"PERSON_PRESENT": "YES", "POURING": "YES", "REASON": "..."}).
    Raises RuntimeError if llama-mtmd-cli fails or produces no parseable output.
    """
    result = subprocess.run(
        [
            "llama-mtmd-cli",
            "-m", MODEL_PATH,
            "--mmproj", MMPROJ_PATH,
            "--image", image_path,
            "-f", prompt_path,
            "-n", str(n_predict),
            "--temp", "0",
        ],
        capture_output=True,
        text=True,
        timeout=120,
    )
    if result.returncode != 0:
        raise RuntimeError(f"llama-mtmd-cli failed (exit {result.returncode}): {result.stderr[-2000:]}")

    fields = {}
    for line in result.stdout.splitlines():
        m = _KV_LINE.match(line.strip())
        if m:
            fields[m.group(1)] = m.group(2)

    if not fields:
        raise RuntimeError(f"No parseable KEY: VALUE output from model. Raw stdout:\n{result.stdout}")

    return fields

File: backend/inference.go
package main

import (
	"bytes"
	"context"
	"fmt"
	"os/exec"
	"regexp"
	"strings"
	"time"
)

var kvLine = regexp.MustCompile(`^([A-Z_]+):\s*(.*?)\s*$`)

// runVLM runs the InternVL GGUF model on one image with one prompt file via
// llama.cpp, returning the KEY: VALUE fields it printed.
func runVLM(cfg *Config, imagePath, promptPath string) (map[string]string, error) {
	ctx, cancel := context.WithTimeout(context.Background(), 120*time.Second)
	defer cancel()

	cmd := exec.CommandContext(ctx, "llama-mtmd-cli",
		"-m", cfg.Model.Path,
		"--mmproj", cfg.Model.MmprojPath,
		"--image", imagePath,
		"-f", promptPath,
		"-n", "150",
		"--temp", "0",
	)

	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	if err := cmd.Run(); err != nil {
		return nil, fmt.Errorf("llama-mtmd-cli failed: %w: %s", err, lastN(stderr.String(), 2000))
	}

	fields := map[string]string{}
	for _, line := range strings.Split(stdout.String(), "\n") {
		if m := kvLine.FindStringSubmatch(strings.TrimSpace(line)); m != nil {
			fields[m[1]] = m[2]
		}
	}
	if len(fields) == 0 {
		return nil, fmt.Errorf("no parseable KEY: VALUE output from model, raw stdout: %s", stdout.String())
	}
	return fields, nil
}

func lastN(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[len(s)-n:]
}
